### PR TITLE
add cpu.cfs_period_us  support

### PR DESCRIPTION
**Pull request by Mashimiao**
Signed-off-by: Ma Shimiao mashimiao.fnst@cn.fujitsu.com

We can use this to control cpu period and quota of a container.

### CODE REVIEW

**Comment by cpuguy83** [comment]
CPU quota is being implemented here: #10736

**Comment by Mashimiao** [comment]
Ah, I have searched and did not find...
I will delete code about quota

**Comment by icecrime** [comment]
Ping @crosbymichael @LK4D4: WDYT?

File: runconfig/hostconfig.go [hunk 1]
```diff
@@ -176,6 +180,7 @@ func ContainerHostConfigFromJob(job *engine.Job) *HostConfig {
 		Memory:          job.GetenvInt64("Memory"),
 		MemorySwap:      job.GetenvInt64("MemorySwap"),
 		CpuShares:       job.GetenvInt64("CpuShares"),
+		CpuPeriod:       job.GetenvInt64("CpuPeriod"),
```
Comment (hqhq) [inline runconfig/hostconfig.go]: Hi,

As I commented in https://github.com/docker/docker/pull/11034#discussion-diff-27010644, we don't actually need any change in `ContainerHostConfigFromJob()`.

File: docs/sources/reference/api/docker_remote_api_v1.19.md [hunk 2]
```diff
@@ -341,6 +344,8 @@ Return low-level information on the container `id`
 			"ContainerIDFile": "",
 			"CpusetCpus": "",
 			"CpuShares": 0,
+			"CpuPeriod": 100000,
+			"CpuQuota": -1,
```
Comment (hqhq) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: Still got Quota code?

File: daemon/execdriver/driver.go [hunk 3]
```diff
@@ -109,6 +109,7 @@ type Resources struct {
 	Memory     int64            `json:"memory"`
 	MemorySwap int64            `json:"memory_swap"`
 	CpuShares  int64            `json:"cpu_shares"`
+	CpuPeriod  int64            `json:cpu_period`
```
Comment (hqhq) [inline daemon/execdriver/driver.go]: Missing quotation marks?

File: runconfig/hostconfig.go (hunk 1 above)
Comment (Mashimiao) [inline runconfig/hostconfig.go]: Ah, @hqhq. Thanks, you'r right.

File: docs/sources/reference/api/docker_remote_api_v1.19.md (hunk 2 above)
Comment (Mashimiao) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: Thanks, this is a mistake.

File: daemon/execdriver/driver.go (hunk 3 above)
Comment (Mashimiao) [inline daemon/execdriver/driver.go]: Thanks for reviewing, @hqhq . fixed.

File: integration-cli/docker_cli_run_test.go [hunk 4]
```diff
@@ -1362,6 +1362,17 @@ func TestRunProcWritableInPrivilegedContainers(t *testing.T) {
 	logDone("run - proc writable in privileged container")
 }
 
+func TestRunWithCpuPeriod(t *testing.T) {
+	defer deleteAllContainers()
+
+	cmd := exec.Command(dockerBinary, "run", "--cpu-period", "200000", "busybox", "true")
+	if code, err := runCommand(cmd); err != nil || code != 0 {
+		t.Fatalf("container should run successfuly with cpu-period of 200000: %s", err)
```
Comment (hqhq) [inline integration-cli/docker_cli_run_test.go]: s/successfuly/successfully

**Comment by hqhq** [comment]
And you need doc update for `docs/sources/reference/run.md`

File: integration-cli/docker_cli_run_test.go (hunk 4 above)
Comment (Mashimiao) [inline integration-cli/docker_cli_run_test.go]: Thanks. fixed.

**Comment by Mashimiao** [comment]
updated doc

**Comment by Mashimiao** [comment]
Ping @crosbymichael @LK4D4

**Comment by Mashimiao** [comment]
ping @jfrazelle , @cpuguy83, @crosbymichael , @LK4D4 
according to #10736 , updated PR

**Comment by Mashimiao** [comment]
rebased.

**Comment by calavera** [comment]
Code LGTM.

/cc @icecrime

File: pkg/sysinfo/sysinfo.go [hunk 5]
```diff
@@ -45,8 +46,13 @@ func New(quiet bool) *SysInfo {
 			logrus.Warnf("%v", err)
 		}
 	} else {
-		_, err1 := ioutil.ReadFile(path.Join(cgroupCpuMountpoint, "cpu.cfs_quota_us"))
-		sysInfo.CpuCfsQuota = err1 == nil
+		_, err := ioutil.ReadFile(path.Join(cgroupCpuMountpoint, "cpu.cfs_period_us"))
+		sysInfo.CpuCfsPeriod = err == nil
+		if !sysInfo.CpuCfsPeriod && !quiet {
+			logrus.Warnf("WARING: Your kernel does not support cgroup cfs period")
```
Comment (coolljt0725) [inline pkg/sysinfo/sysinfo.go]: the `WARING:` is redundant

Comment (Mashimiao) [inline pkg/sysinfo/sysinfo.go]: redundant?

Comment (coolljt0725) [inline pkg/sysinfo/sysinfo.go]: `logrus.Warf` will has a `WARN`, I think this is no need `WARNING` in your message

Comment (Mashimiao) [inline pkg/sysinfo/sysinfo.go]: Thanks! fixed

**Comment by mrunalp** [comment]
:+1:

**Comment by Mashimiao** [comment]
rebased and ping

**Comment by LK4D4** [comment]
LGTM

File: docs/sources/reference/api/docker_remote_api_v1.19.md [hunk 6]
```diff
@@ -191,6 +192,7 @@ Json Parameters:
       always use this with `memory`, and make the value larger than `memory`.
 -   **CpuShares** - An integer value containing the CPU Shares for container
       (ie. the relative weight vs other containers).
+-   **CpuPeriod** - The length of a period (in microseconds).
```
Comment (SvenDowideit) [inline docs/sources/reference/api/docker_remote_api_v1.19.md]: length of which period?

File: docs/sources/reference/run.md [hunk 7]
```diff
@@ -587,6 +588,21 @@ division of CPU shares:
     101    {C1}		1	100% of CPU1
     102    {C1}		2	100% of CPU2
 
+### CPU period constraint
+
+The default period of CPUs is 100ms. We can use `--cpu-period` to set the
```
Comment (SvenDowideit) [inline docs/sources/reference/run.md]: I don't think `The default period of CPUs is 100ms.` is right - isn't it a CFS scheduler default period - used by Linux..

ie, its not a CPU feature, but a Linux scheduler feature.

File: docs/sources/reference/run.md [hunk 8]
```diff
@@ -587,6 +588,21 @@ division of CPU shares:
...
+The default period of CPUs is 100ms. We can use `--cpu-period` to set the 
+period of CPUs to limit containers's CPU usage. And usually `--cpu-period`
+should work with `--cpu-quota`.
+
+Examples:
+
+    $ docker run -ti --cpu-period=100000 --cpu-quota=50000 ubuntu:14.04 /bin/bash
+
+If there is 1 CPU, this means the container can get 50% worth run-time of 1 CPU
+every 100ms.
```
Comment (SvenDowideit) [inline docs/sources/reference/run.md]: mention that the `--cpu-period=100000` is the default, or use a non-default and explain it?

Comment (Mashimiao) [inline docs/sources/reference/run.md]: fixed.

File: docs/sources/reference/run.md (hunk 7 above)
Comment (Mashimiao) [inline docs/sources/reference/run.md]: fixed.

**Comment by Mashimiao** [comment]
ping @SvenDowideit

**Comment by calavera** [comment]
@moxiegirl, @thaJeztah, @jamtur01 can anyone of you take a look to the docs here?

File: docs/sources/reference/run.md [hunk 9]
```diff
@@ -616,6 +617,21 @@ division of CPU shares:
     101    {C1}		1	100% of CPU1
     102    {C1}		2	100% of CPU2
 
+### CPU period constraint
+
+The default CPU CFS (Completely Fair Scheduler) period is 100ms. We can use
+`--cpu-period` to set the period of CPUs to limit containers's CPU usage.
```
Comment (jamtur01) [inline docs/sources/reference/run.md]: to limit the container's ...

**Comment by jamtur01** [comment]
One minor comment otherwise LGTM

File: docs/sources/reference/run.md [hunk 10]
```diff
@@ -616,6 +617,21 @@ division of CPU shares:
...
+
+The default CPU CFS (Completely Fair Scheduler) period is 100ms. We can use
+`--cpu-period` to set the period of CPUs to limit containers's CPU usage. 
+And usually `--cpu-period` should work with `--cpu-quota`.
+
+Examples:
+
+    $ docker run -ti --cpu-period=50000 --cpu-quota=25000 ubuntu:14.04 /bin/bash
+
+If there is 1 CPU, this means the container can get 50% worth run-time of 1 CPU
```
Comment (thaJeztah) [inline docs/sources/reference/run.md]: 50% worth *of* run-time?

**Comment by thaJeztah** [comment]
One more minor nit, otherwise LGTM

@calavera Both doc comments are really minor, so I'm not really picky if it would be merged and patched later (or carried) to speed up the merge :)

File: docs/sources/reference/run.md (hunk 9 above)
Comment (Mashimiao) [inline docs/sources/reference/run.md]: fixed

File: docs/sources/reference/run.md (hunk 10 above)
Comment (Mashimiao) [inline docs/sources/reference/run.md]: fixed

**Comment by thaJeztah** [comment]
Thanks @Mashimiao 

LGTM (if the lights go green)

**Comment by jamtur01** [comment]
Docs LGTM

**Pull request merged**